core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -267,15 +270,21 @@ public boolean isVersionActual(int segment, EntryVersion version) {
    }
 
    @Override
-   public void notifyKeyTransferFinished(int segment, boolean expectValues) {
-      SegmentState update = expectValues ? SegmentState.VALUE_TRANSFER : SegmentState.OWNED;
-      SegmentState previous;
-      do {
-         previous = segmentStates.get(segment);
-         // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
-         // when the CONFIRM_REVOKED_SEGMENTS failed.
-         log.tracef("Finishing transfer for segment %d = %s -> %s", segment, previous, update);
-      } while (!segmentStates.compareAndSet(segment, previous, update));
+   public void notifyKeyTransferFinished(int segment, boolean expectValues, boolean cancelled) {
+      SegmentState update;
+      if (expectValues) {
+         update = SegmentState.VALUE_TRANSFER;
+      } else if (cancelled) {
+         // The transfer is cancelled when a newer topology is being installed.
+         update = SegmentState.NOT_OWNED;
+      } else {
+         update = SegmentState.OWNED;
+      }
+      // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
+      // when the CONFIRM_REVOKED_SEGMENTS failed.
+      // Another possibility is when there is already another rebalance starting, then it might be NOT_OWNED
```

Comment:
When two nodes crash in sequence, the other rebalance can be sent out before all nodes confirm the previous rebalance. If both crashed were coords (at their time), you can imagine that keeping track of who finished would be problematic - that's why we support canceling the state transfer even in dist/repl caches.